### PR TITLE
Synchronize the CUDA stream before returning from execute

### PR DESCRIPTION
**Pull request by shoumikhin**
## The problem

The CUDA delegate finishes its work on the GPU **asynchronously**, then returns without waiting for it to finish. GPU work is queued on a "stream" (an ordered queue of operations), and work on one stream is not visible to a reader on a different stream unless the producer waits for it first.

So whoever runs next can read an output buffer that is still being written. The result is wrong numbers rather than an error, and the same wrong numbers on every run, which makes it look like a modeling bug rather than a race.

The delegate's own helper already documents the rule being broken:

```
// NOTE: In the fast path the copy is asynchronous. The caller must synchronize
// the stream before consuming the ETensor data.
```

Nothing on the normal execution path does that.

## Why it is reachable

The normal path launches the kernels and, when AOTInductor hands back its own output buffer, an asynchronous device-to-device copy into the planned buffer. Then it returns.

That is only safe while every consumer happens to run on the same stream. It is not safe when the output is consumed by:

- a delegate from a **different backend**, which selects its own stream and has no event ordering it after this one, or
- a portable kernel on the default stream.

The two CUDA graph paths in the same function already synchronize before returning (replay and capture). Only the normal path does not, so this change makes it consistent with its neighbours.

## The fix

One synchronize at the end of the normal path, mirroring what the graph paths already do:

```cpp
    ET_CUDA_CHECK_OR_RETURN_ERROR(cudaStreamSynchronize(cuda_stream));

    return Error::Ok;
```

Nine added lines, no deletions, one file.

## Testing

Found while running a program that interleaves this backend with another backend's delegate in one method. The failure scales with the number of handoffs between the two: with a couple of handoffs the timing happened to work out, and with more it was consistently wrong.

The check that isolates it is an A/B on one unchanged program, where the only difference is whether the GPU is forced to run everything in order:

```
default                    first token 13377, then 0 0 0 0 0 0 0 0   (dead)
CUDA_LAUNCH_BLOCKING=1     first token  3854, then 13 13 17 13 ...   (alive)
with this change           first token  3854, then 13 13 17 13 ...   (identical)
```

Serialising all GPU work and adding the missing synchronize produce **byte-identical** output, which is what a correctly fixed ordering hazard looks like. Also confirmed this does not disturb a program that was already correct: a longer-sequence variant of the same model still produces its previous, verified output.

## A note on a regression test

I did not include one, and I would rather say why than ship something that does not actually test this.

A genuine test needs a consumer reading on a **different stream**, which in practice means a second backend in the same method. I could not find an existing test that runs two different delegates in one method and checks numerics, so there is no natural place to extend.

I tried two versions and discarded both:

1. A C++ unit test asserting cross-stream visibility. It verified CUDA's own documented semantics, so it passed with and without the change. A tautology, not a regression test.
2. A Python test lowering a multi-output model to this backend and comparing against eager, following the pattern of the existing `test_output_stride_rearrange.py`. It crashed in the pybindings loader in my environment, and so did that pre-existing test unchanged, so I could not validate any result from it.

Happy to add a test if a maintainer can point me at the right harness for a two-backend numeric check.

### CODE REVIEW

**Comment by shoumikhin** [comment]
Closing this. I could not reproduce the effect I claimed, and I would rather retract it than leave a maintainer to discover that.

## What went wrong with my evidence

I reported that adding the synchronize changed the output of a program from dead to alive, and that the result was byte-identical to running under `CUDA_LAUNCH_BLOCKING=1`.

That comparison was invalid. The "patched" library I measured was a partial rebuild I had produced myself (about 200 KB, missing the device code section) while the "unpatched" one was the real full build (about 7 MB). So I was comparing two different builds and attributing the difference to this one line.

## The correct experiment

I built two libraries of the **same** build type that differ only by this change, one with 7 synchronize call sites and one with 8, and ran the same programs with nothing else changed:

```
program A (single-token variant)
  7 sites (without this change):  dead output, and it varies between runs
  8 sites (with this change):     dead output, and it varies between runs

program B (16-token variant)
  7 sites:  first token 16733, then zeros
  8 sites:  first token 16733, then zeros   (identical)
```

The change makes no observable difference on either program. Whatever is wrong in my setup, this is not it.

## What still seems worth a look, separately from this change

I am not claiming these are bugs, only that they looked odd while I was in the area, and someone who knows this code will judge them faster than I can:

1. `copy_slimtensor_to_device_etensor_async` in `backends/cuda/runtime/utils.h` documents that "the caller must synchronize the stream before consuming the ETensor data". Its only caller is the normal path in `cuda_backend.cpp`, which does not synchronize, while the CUDA graph replay and capture paths in the same function do. That asymmetry may well be intentional, and my failed experiment is not evidence either way.

2. Separately, the output I was chasing turned out to be a **partially written** buffer: a contiguous prefix of correct-length logits followed by exact zeros, with the boundary moving between runs. That is a real symptom of something in my own setup, and I will track it down before proposing anything else here.

Sorry for the noise. Thanks for the earlier question on the JetPack version, which was on a different change and still stands.
